File: apps/api/src/analytics/diagnostics.ts
import { Effect, Schema, Clock } from 'effect';
import { Infrastructure } from '../platform/resources';
import { attempt, invalid, ApiError, attemptSync } from '../shared/errors';
import { decode, Id } from '../shared/validation';
import { hash, settings } from './tracking';
import { allowance } from '../billing/service';
const Input = Schema.Struct({
  siteId: Id,
  environmentId: Id,
  id: Id,
  pageId: Id,
  url: Schema.String.check(Schema.isMaxLength(2048)),
  kind: Schema.Literals(['error', 'vital']),
  payload: Schema.Unknown,
  consent: Schema.Boolean,
});
export function sanitize(value: string, max: number) {
  return (
    value
      // eslint-disable-next-line no-control-regex -- Reject control characters at the privacy boundary.
      .replace(/[\x00-\x09\x0b-\x1f\x7f]/g, '')
      .slice(0, max)
      .replace(/https?:\/\/[^\s)"']+/g, (value) => {
        try {
          const url = new URL(value);
          url.search = '';
          url.hash = '';
          url.username = '';
          url.password = '';
          return url.toString();
        } catch {
          return '';
        }
      })
      .replace(/[a-z0-9._%+-]+@[a-z0-9.-]+\.[a-z]{2,}/gi, '[redacted]')
      .replace(/["'][^"'\n]*["']|\b[A-Za-z0-9_-]{32,}\b/g, '[redacted]')
      .slice(0, max)
  );
}
function payload(kind: string, value: unknown) {
  if (kind === 'vital') {
    const input = decode(
      Schema.Struct({
        name: Schema.Literals(['CLS', 'LCP', 'INP']),
        value: Schema.Number.check(
          Schema.isGreaterThanOrEqualTo(0),
          Schema.isLessThanOrEqualTo(1800000),
        ),
      }),
      value,
    );
    if (input.name === 'CLS' && input.value > 100) throw invalid('Invalid Web Vital.');
    return input;
  }
  const input = decode(
      Schema.Struct({
        message: Schema.String,
        source: Schema.optional(Schema.String),
        stack: Schema.optional(Schema.String),
        line: Schema.optional(Schema.Number.check(Schema.isInt())),
        column: Schema.optional(Schema.Number.check(Schema.isInt())),
      }),
      value,
    ),
    message = sanitize(input.message, 500);
  if (!message.trim()) throw invalid('An error needs a message.');
  return {
    message,
    source: sanitize(input.source ?? '', 512),
    stack: sanitize(input.stack ?? '', 2000),
    line: Math.max(0, Math.min(1000000, input.line ?? 0)),
    column: Math.max(0, Math.min(1000000, input.column ?? 0)),
  };
}
export type Diagnostic = {
  site: string;
  env: string;
  id: string;
  page: string;
  at: string;
  kind: string;
  path: string;
  device: string;
  visitor: string;
  fingerprint: string;
  payload: ReturnType<typeof payload>;
  consent: boolean;
};
export const telemetry = Effect.fn('telemetry')(function* (
  body: unknown,
  headers: Headers,
  ip: string,
) {
  const r = yield* Infrastructure;
  const input = yield* attemptSync(() => decode(Input, body)),
    ua = (headers.get('user-agent') ?? '').slice(0, 512);
  if (headers.get('dnt') === '1' || /bot|crawler|spider|headless|lighthouse|curl|wget/i.test(ua))
    return { accepted: false };
  return yield* attempt(async () => {
    const secret = process.env.VISITOR_HASH_SECRET ?? r.config.BETTER_AUTH_SECRET;
    const count = Number(
      await r.redis.send('EVAL', [
        "local n=redis.call('INCR',KEYS[1]);if n==1 then redis.call('EXPIRE',KEYS[1],60) end;return n",
        '1',
        `${r.config.queuePrefix}:diagnostics:${hash(secret, ip)}`,
      ]),
    );
    if (count > 60)
      throw new ApiError({
        status: 429,
        code: 'rate_limited',
        message: 'Too many requests. Try again in a minute.',
      });
    const [env] =
      await r.primary`SELECT e.*,s.owner_id FROM environments e JOIN sites s ON s.id=e.site_id WHERE e.id=${input.environmentId}::uuid AND s.id=${input.siteId}::uuid`;
    if (!env) throw invalid('Environment not found.');
    if (
      !env.enabled ||
      !(
        env.feature_settings[input.kind === 'error' ? 'errors' : 'webVitals'] ??
        input.kind === 'vital'
      ) ||
      !(await allowance(r.primary, env.owner_id))
    )
      return { accepted: false };
    const page = new URL(input.url),
      local = ['localhost', '127.0.0.1', '[::1]'].includes(page.hostname);
    if (
      !['http:', 'https:'].includes(page.protocol) ||
      page.username ||
      page.password ||
      page.origin !== headers.get('origin') ||
      (page.hostname !== env.domain && !(local && env.allow_localhost))
    )
      throw invalid('Invalid website origin.');
    const data = payload(input.kind, input.payload),
      at = new Date().toISOString(),
      fields = data as Record<string, unknown>;
    const diagnostic: Diagnostic = {
      site: input.siteId,
      env: input.environmentId,
      id: input.id,
      page: input.pageId,
      at,
      kind: input.kind,
      path: page.pathname,
      device: settings(env.tracking_settings).device
        ? /iPad|Tablet/i.test(ua)
          ? 'tablet'
          : /Mobile|iPhone|Android/i.test(ua)
            ? 'mobile'
            : 'desktop'
        : '',
      visitor: hash(secret, [env.id, at.slice(0, 10), ip, ua]),
      fingerprint: hash(secret, [
        env.id,
        fields.name ?? null,
        fields.message ?? null,
        fields.source ?? null,
        fields.line ?? null,
      ]),
      payload: data,
      consent: input.consent,
    };
    await r.queue.add('diagnostic', { diagnostic });
    return { accepted: true };
  });
});
export const ingestDiagnostic = Effect.fn('ingestDiagnostic')(function* (d: Diagnostic) {
  const r = yield* Infrastructure;
  const now = yield* Clock.currentTimeMillis;
  if (Date.parse(d.at) < now - 30 * 86400000 || Date.parse(d.at) > now + 60000) return;
  const clean = payload(d.kind, d.payload);
  yield* attempt(() =>
    r.primary.begin(async (tx) => {
      const [owner] =
        await tx`SELECT s.owner_id FROM sites s JOIN environments e ON e.site_id=s.id WHERE s.id=${d.site}::uuid AND e.id=${d.env}::uuid`;
      if (!owner) return;
      await tx`SELECT id FROM "user" WHERE id=${owner.owner_id} FOR UPDATE`;
      const active = await allowance(tx, owner.owner_id);
      if (!active) return;
      const sites =
        await tx`SELECT id FROM sites WHERE owner_id=${owner.owner_id} ORDER BY created_at,id`;
      if (
        active.websiteLimit !== null &&
        sites.findIndex((s: { id: string }) => s.id === d.site) >= active.websiteLimit
      )
        return;
      const feature = d.kind === 'error' ? 'errors' : 'webVitals';
      const [row] =
        await tx`SELECT e.enabled AND s.enabled AND coalesce((e.feature_settings->>${feature})::boolean,${feature}='webVitals') AND (e.tracking_mode='cookieless' OR ${d.consent}) AND NOT EXISTS(SELECT 1 FROM user_suspensions WHERE user_id=s.owner_id) AND NOT EXISTS(SELECT 1 FROM site_suspensions WHERE site_id=s.id) AS allowed FROM environments e JOIN sites s ON s.id=e.site_id WHERE e.id=${d.env}::uuid AND s.id=${d.site}::uuid FOR KEY SHARE OF s,e`;
      if (!row?.allowed) return;
      const previous =
        await tx`SELECT * FROM diagnostic_events WHERE environment_id=${d.env}::uuid AND id=${d.id}::uuid ORDER BY received_at DESC LIMIT 1`;
      if (previous.length) {
        const old = previous[0];
        if (
          old.kind !== 'vital' ||
          d.kind !== 'vital' ||
          old.page_id !== d.page ||
          old.payload.name !== ('name' in clean ? clean.name : undefined) ||
          new Date(old.received_at).getTime() >= Date.parse(d.at)
        )
          return;
        await tx`DELETE FROM diagnostic_events WHERE environment_id=${d.env}::uuid AND id=${d.id}::uuid AND received_at=${old.received_at}`;
      }
      await tx`INSERT INTO diagnostic_events(environment_id,id,page_id,received_at,kind,path,device,visitor,fingerprint,payload) VALUES(${d.env}::uuid,${d.id}::uuid,${d.page}::uuid,${d.at},${d.kind},${d.path},${d.device},${d.visitor},${d.fingerprint},${JSON.stringify(clean)}::text::jsonb) ON CONFLICT DO NOTHING`;
    }),
  );
});

File: crates/api/src/diagnostics.rs
use crate::{AppState, billing, error::ApiError, http::json_body, sites::identifier, tracking};
use axum::{
    Json, Router,
    extract::{Request, State},
    http::{HeaderMap, StatusCode},
    routing::post,
};
use chrono::{DateTime, Duration, Utc};
use regex::Regex;
use serde::{Deserialize, Serialize};
use serde_json::{Value, json};
use sqlx::Row;
use std::sync::LazyLock;
use uuid::Uuid;

#[derive(Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
struct Input {
    site_id: String,
    environment_id: String,
    id: String,
    page_id: String,
    url: String,
    kind: String,
    payload: Value,
    consent: bool,
}

#[derive(Deserialize, Serialize)]
pub struct Diagnostic {
    site: Uuid,
    env: Uuid,
    id: Uuid,
    page: Uuid,
    at: DateTime<Utc>,
    kind: String,
    path: String,
    device: String,
    visitor: String,
    fingerprint: String,
    payload: Value,
    consent: bool,
}

fn truncate(value: &str, max: usize) -> String {
    let mut size = 0;
    value
        .chars()
        .take_while(|c| {
            size += c.len_utf16();
            size <= max
        })
        .collect()
}

fn sanitize(value: &str, max: usize) -> String {
    static URL: LazyLock<Regex> =
        LazyLock::new(|| Regex::new(r#"https?://[^\s)"']+"#).expect("constant regex"));
    static EMAIL: LazyLock<Regex> = LazyLock::new(|| {
        Regex::new(r"(?i)[a-z0-9._%+-]+@[a-z0-9.-]+\.[a-z]{2,}").expect("constant regex")
    });
    static SECRETS: LazyLock<Regex> = LazyLock::new(|| {
        Regex::new(r#"["'][^"'\n]*["']|\b[A-Za-z0-9_-]{32,}\b"#).expect("constant regex")
    });
    let clean: String = value
        .chars()
        .filter(|c| !matches!(*c,'\x00'..='\x09'|'\x0b'..='\x1f'|'\x7f'))
        .collect();
    let short = truncate(&clean, max);
    let urls = URL.replace_all(&short, |captures: &regex::Captures<'_>| {
        url::Url::parse(&captures[0])
            .map(|mut url| {
                url.set_query(None);
                url.set_fragment(None);
                let _ = url.set_username("");
                let _ = url.set_password(None);
                url.to_string()
            })
            .unwrap_or_default()
    });
    let emails = EMAIL.replace_all(&urls, "[redacted]");
    truncate(&SECRETS.replace_all(&emails, "[redacted]"), max)
}

fn payload(kind: &str, value: Value) -> Result<Value, ApiError> {
    if kind == "vital" {
        #[derive(Deserialize, Serialize)]
        #[serde(deny_unknown_fields)]
        struct Vital {
            name: String,
            value: f64,
        }
        let input: Vital = serde_json::from_value(value).map_err(|_| ApiError::invalid())?;
        if !["CLS", "LCP", "INP"].contains(&input.name.as_str())
            || !input.value.is_finite()
            || !(0.0..=1_800_000.0).contains(&input.value)
            || (input.name == "CLS" && input.value > 100.0)
        {
            return Err(ApiError::invalid());
        }
        return Ok(json!(input));
    }
    if kind != "error" {
        return Err(ApiError::invalid());
    }
    #[derive(Deserialize)]
    #[serde(deny_unknown_fields)]
    struct Error {
        message: String,
        source: Option<String>,
        stack: Option<String>,
        line: Option<i64>,
        column: Option<i64>,
    }
    let input: Error = serde_json::from_value(value).map_err(|_| ApiError::invalid())?;
    let message = sanitize(&input.message, 500);
    if message.trim().is_empty() {
        return Err(ApiError::new(
            StatusCode::BAD_REQUEST,
            "invalid_request",
            "An error needs a message.",
        ));
    }
    Ok(
        json!({"message":message,"source":sanitize(input.source.as_deref().unwrap_or(""),512),"stack":sanitize(input.stack.as_deref().unwrap_or(""),2000),"line":input.line.unwrap_or(0).clamp(0,1_000_000),"column":input.column.unwrap_or(0).clamp(0,1_000_000)}),
    )
}

pub fn routes() -> Router<AppState> {
    Router::new().route("/api/telemetry", post(telemetry_route))
}
async fn telemetry_route(
    State(state): State<AppState>,
    request: Request,
) -> Result<(StatusCode, Json<Value>), ApiError> {
    let headers = request.headers().clone();
    Ok((
        StatusCode::ACCEPTED,
        Json(telemetry(&state, json_body(request).await?, &headers).await?),
    ))
}

async fn telemetry(state: &AppState, input: Input, headers: &HeaderMap) -> Result<Value, ApiError> {
    let site = identifier(&input.site_id)?;
    let environment = identifier(&input.environment_id)?;
    let id = identifier(&input.id)?;
    let page_id = identifier(&input.page_id)?;
    if !matches!(input.kind.as_str(), "error" | "vital") || input.url.encode_utf16().count() > 2048
    {
        return Err(ApiError::invalid());
    }
    if tracking::excluded(headers) {
        return Ok(json!({"accepted":false}));
    }
    let ip = tracking::client_ip(headers);
    tracking::rate_limit(state, "diagnostics", ip, 60).await?;
    let env=sqlx::query("SELECT e.*,s.owner_id FROM environments e JOIN sites s ON s.id=e.site_id WHERE e.id=$1 AND s.id=$2")
        .bind(environment).bind(site).fetch_optional(&state.pool).await?.ok_or_else(||ApiError::new(StatusCode::BAD_REQUEST,"invalid_request","Environment not found."))?;
    let features: Value = env.get("feature_settings");
    let mut connection = state.pool.acquire().await?;
    if !env.get::<bool, _>("enabled")
        || !features[if input.kind == "error" {
            "errors"
        } else {
            "webVitals"
        }]
        .as_bool()
        .unwrap_or(input.kind == "vital")
        || billing::allowance(
            &mut connection,
            env.get("owner_id"),
            state.billing.organization_id(),
            state.config.external_effects,
        )
        .await?
        .is_none()
    {
        return Ok(json!({"accepted":false}));
    }
    drop(connection);
    let (page, _) = tracking::page_url(
        &input.url,
        tracking::origin(headers),
        env.get("domain"),
        env.get("allow_localhost"),
    )?;
    let data = payload(&input.kind, input.payload)?;
    let at = Utc::now();
    let ua = tracking::user_agent(headers);
    let diagnostic = Diagnostic {
        site,
        env: environment,
        id,
        page: page_id,
        at,
        kind: input.kind,
        path: page.path().into(),
        device: if tracking::settings(&env.get::<Value, _>("tracking_settings"))["device"] == true {
            tracking::device(&ua).into()
        } else {
            String::new()
        },
        visitor: tracking::hash(
            &state.config.visitor_secret,
            &json!([environment, at.date_naive(), ip, ua]),
        ),
        fingerprint: tracking::hash(
            &state.config.visitor_secret,
            &json!([
                environment,
                data["name"],
                data["message"],
                data["source"],
                data["line"]
            ]),
        ),
        payload: data,
        consent: input.consent,
    };
    state.queue.enqueue(&diagnostic).await?;
    Ok(json!({"accepted":true}))
}

pub async fn ingest(state: &AppState, d: Diagnostic) -> Result<(), ApiError> {
    let now = Utc::now();
    if d.at < now - Duration::days(30) || d.at > now + Duration::seconds(60) {
        return Ok(());
    }
    let clean = payload(&d.kind, d.payload)?;
    let mut tx = state.pool.begin().await?;
    let owner:Option<String>=sqlx::query_scalar("SELECT s.owner_id FROM sites s JOIN environments e ON e.site_id=s.id WHERE s.id=$1 AND e.id=$2")
        .bind(d.site).bind(d.env).fetch_optional(&mut *tx).await?;
    let Some(owner) = owner else {
        return Ok(());
    };
    sqlx::query("SELECT id FROM public.\"user\" WHERE id=$1 FOR UPDATE")
        .bind(&owner)
        .execute(&mut *tx)
        .await?;
    let Some(active) = billing::allowance(
        &mut tx,
        &owner,
        state.billing.organization_id(),
        state.config.external_effects,
    )
    .await?
    else {
        return Ok(());
    };
    let sites: Vec<Uuid> =
        sqlx::query_scalar("SELECT id FROM sites WHERE owner_id=$1 ORDER BY created_at,id")
            .bind(&owner)
            .fetch_all(&mut *tx)
            .await?;
    if active.entitlement.website_limit.is_some_and(|limit| {
        sites
            .iter()
            .position(|id| *id == d.site)
            .is_none_or(|i| i as i64 >= limit)
    }) {
        return Ok(());
    }
    let feature = if d.kind == "error" {
        "errors"
    } else {
        "webVitals"
    };
    let allowed:Option<bool>=sqlx::query_scalar("SELECT e.enabled AND s.enabled AND coalesce((e.feature_settings->>$1)::boolean,$1='webVitals') AND (e.tracking_mode='cookieless' OR $2) AND NOT EXISTS(SELECT 1 FROM user_suspensions WHERE user_id=s.owner_id) AND NOT EXISTS(SELECT 1 FROM site_suspensions WHERE site_id=s.id) AS allowed FROM environments e JOIN sites s ON s.id=e.site_id WHERE e.id=$3 AND s.id=$4 FOR KEY SHARE OF s,e")
        .bind(feature).bind(d.consent).bind(d.env).bind(d.site).fetch_optional(&mut *tx).await?;
    if allowed != Some(true) {
        return Ok(());
    }
    let previous=sqlx::query("SELECT * FROM diagnostic_events WHERE environment_id=$1 AND id=$2 ORDER BY received_at DESC LIMIT 1")
        .bind(d.env).bind(d.id).fetch_optional(&mut *tx).await?;
    if let Some(old) = previous {
        if old.get::<String, _>("kind") != "vital"
            || d.kind != "vital"
            || old.get::<Uuid, _>("page_id") != d.page
            || old.get::<Value, _>("payload")["name"] != clean["name"]
            || old.get::<DateTime<Utc>, _>("received_at") >= d.at
        {
            return Ok(());
        }
        sqlx::query(
            "DELETE FROM diagnostic_events WHERE environment_id=$1 AND id=$2 AND received_at=$3",
        )
        .bind(d.env)
        .bind(d.id)
        .bind(old.get::<DateTime<Utc>, _>("received_at"))
        .execute(&mut *tx)
        .await?;
    }
    sqlx::query("INSERT INTO diagnostic_events(environment_id,id,page_id,received_at,kind,path,device,visitor,fingerprint,payload) VALUES($1,$2,$3,$4,$5,$6,$7,$8,$9,$10) ON CONFLICT DO NOTHING")
        .bind(d.env).bind(d.id).bind(d.page).bind(d.at).bind(d.kind).bind(d.path).bind(d.device).bind(d.visitor).bind(d.fingerprint).bind(clean).execute(&mut *tx).await?;
    tx.commit().await?;
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn telemetry_redacts_private_values_and_bounds_vitals() {
        let clean=payload("error",json!({"message":"failed at https://user:password@example.test/a?token=secret#hash for Alice@example.test 'private text' abcdefghijklmnopqrstuvwxyz0123456789","line":-10,"column":2_000_000})).unwrap();
        let message = clean["message"].as_str().unwrap();
        assert!(message.contains("https://example.test/a"));
        for private in ["password", "token", "Alice@", "private text", "abcdef"] {
            assert!(!message.contains(private));
        }
        assert_eq!(clean["line"], 0);
        assert_eq!(clean["column"], 1_000_000);
        assert!(payload("vital", json!({"name":"CLS","value":100.1})).is_err());
        assert!(payload("vital", json!({"name":"LCP","value":10})).is_ok());
    }
}
